Automattic/jetpack
# Implement pre-connection JITMs

#### Changes proposed in this Pull Request:
Implement pre-connection messages (aka JITMs) aimed to introduce users to Jetpack's features and encourage them to connect.

This project has 2 parts:
1) Refactor the JITM class (class-jitm.php) leaving common methods in it and moving post-connection logic to a new subclass Post_Connection_JITM (class-post-connection-jitm.php). No changes have been made to the existing logic.
2) Create a Pre_connection_JITM subclass (class-pre-connection-jitm.php) with new pre-connection specific messages and methods.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This is a new feature, mostly discussed in pbtFPC-nZ-p2 and related posts.

#### Testing instructions:
1) Get a brand new Jetpack environment, 
2) Activate the Jetpack plugin but DO NOT connect it.
3) Somewhere like in wp-config.php, add the following filter:
`add_filter( 'jetpack_pre_connection_prompt_helpers', '__return_true' );`
4) There are 3 messages implemented for this release as specified in pbtFPC-nZ-p2. They should be displayed at edit.php, upload.php, widgets.php like this:

<img width="1168" alt="Screen Shot 2020-05-07 at 20 01 07" src="https://user-images.githubusercontent.com/2166135/81352746-a8886e00-909d-11ea-8a62-3f08350a2415.png">

5) In addition to displaying the JITMs, click the blue "Setup Jetpack" buttons and check it takes you to the connection screen.

#### Proposed changelog entry for your changes:
* Implemented pre-connection JITMs.
